feature/google/src/main/java/com/revenuecat/purchases/google/findBestOffer.kt

```diff
@@ -0,0 +1,108 @@
+@file:Suppress("MaximumLineLength", "MaxLineLength")
+package com.revenuecat.purchases.google
+
+import com.revenuecat.purchases.models.GoogleSubscriptionOption
+import com.revenuecat.purchases.models.PricingPhase
+import com.revenuecat.purchases.models.RecurrenceMode
+
+fun List<GoogleSubscriptionOption>.findBestOffer(): GoogleSubscriptionOption? {
+    val basePlan = this.firstOrNull { it.isBasePlan } ?: return null
+
+    val validOffers = this
+        .filter { !it.isBasePlan }
+        .filter { !it.tags.contains("rc-ignore-best-offer") }
+
+    return findLongestFreeTrial(validOffers) ?: findBestSavingsOffer(basePlan, validOffers) ?: basePlan
+}
+
+private fun findLongestFreeTrial(offers: List<GoogleSubscriptionOption>): GoogleSubscriptionOption? {
+    return offers.mapNotNull { offer ->
+        offer.pricingPhases.filter { pricingPhase ->
+            pricingPhase.priceAmountMicros == 0L
+        }.map {
+            Pair(offer, parseBillPeriodToDays(it.billingPeriod))
+        }.maxByOrNull { it.second }
+    }.maxByOrNull { it.second }?.first
+}
+
+@Suppress("ReturnCount")
+private fun findBestSavingsOffer(
+    basePlan: GoogleSubscriptionOption,
+    offers: List<GoogleSubscriptionOption>
+): GoogleSubscriptionOption? {
+    val basePlanPricingPhase = basePlan.pricingPhases.firstOrNull() ?: return null
+
+    val longestOfferTotalDays = offers
+        .mapNotNull { subscriptionOption ->
+            subscriptionOption.pricingPhases.sumOf { it.totalNumberOfDays() ?: 0 }
+        }.maxByOrNull { it } ?: 0
+
+    if (longestOfferTotalDays < 0) {
+        return null
+    }
+
+    val periodDays = parseBillPeriodToDays(basePlanPricingPhase.billingPeriod)
+    val basePlanCostPerDay = basePlanPricingPhase.priceAmountMicros / periodDays
+
+    return offers.map { offer ->
+        // Maps each phase for total cost up to longest offer duration
+        val pricePerDayPerPhase = offer.pricingPhases
+            .filter { it.recurrenceMode == RecurrenceMode.FINITE_RECURRING }
+            .mapNotNull { pricingPhase ->
+                val periodDays = parseBillPeriodToDays(pricingPhase.billingPeriod)
+
+                if (periodDays > 0) {
+                    val pricePerDay = pricingPhase.priceAmountMicros / periodDays
+                    val totalDaysOfPhase = pricingPhase.totalNumberOfDays() ?: 0
+
+                    val numberOfDaysOnBasePlan = longestOfferTotalDays - totalDaysOfPhase
+
+                    val costOfPricingPhase = totalDaysOfPhase * pricePerDay
+                    val costOfBasePlan = numberOfDaysOnBasePlan * basePlanCostPerDay
+
+                    val totes = costOfPricingPhase + costOfBasePlan
+
+                    Pair(pricingPhase, totes)
```

Comment:
This logic _might_ be wrong for if there is an offer with two pricing phases 🤔 Going to look into how this can be setup in Google Play Console and then write some tests for it